### ISSUE TITLE
Behavior of FreeLine, Horizontal and Vertical movable lines

### ISSUE THREAD

**Issue by elkins0** [CindyJS/CindyJS#79]
"Export to CindyJS..." from Cinderella can produce references to FreeLine, HorizontalLine and VerticalLine. The HTML produced has to be changed to match CindyJS so that the types "HorizontalLine" and "VerticalLine" became "Horizontal" and "Vertical". Even with those changes "Horizontal" and "Vertical" do no work as they reference a non-existant point argument. And while FreeLine displays, it displays randomly and is not movable.

In https://github.com/CindyJS/CindyJS/tree/master/src/js/libgeo/GeoOps.js,
based on `geoOps.Through` and `geoOps.Free` I propose the following replacements for
`geoOps.FreeLine`, `geoOps.Horizontal` and `geoOps.Vertical`. To avoid having to change Cinderella's "Export to CindyJS..."  `geoOps.Horizontal` and `geoOps.Vertical`
could be renamed `geoOps.HorizontalLine` and `geoOps.VerticalLine`. That decision is up to you.

Also, something strange I noticed during my perusal of the code in GeoOps.js. This is the only geoOps.*.putParamToState that has a return in it:
```javascript
geoOps.PointOnLine.putParamToState = function(el, param) {
    return putStateComplexVector(param);
};
```

### geoOps.FreeLine

```javascript
geoOps.FreeLine = {};
geoOps.FreeLine.kind = "L";
geoOps.FreeLine.isMovable = true;
geoOps.FreeLine.initialize = function (el) {
    var pos = geoOps._helper.initializePoint(el);
    putStateComplexVector(pos);
};
geoOps.FreeLine.getParamForInput = function (el, pos, type) {
    var homog = List.cross(pos, List.ez);
    homog = List.cross(homog, pos);
    return List.normalizeMax(homog);
};
geoOps.FreeLine.getParamFromState = function (el) {
    return getStateComplexVector(3);
};
geoOps.FreeLine.putParamToState = function (el, param) {
    putStateComplexVector(param);
};
geoOps.FreeLine.updatePosition = function (el) {
    var param = getStateComplexVector(3);
    putStateComplexVector(param); // copy param
    el.homog = General.withUsage(param, "Line");
};
geoOps.FreeLine.stateSize = 6;
```

### geoOps.Horizontal

```javascript
geoOps.Horizontal = {};
geoOps.Horizontal.kind = "L";
geoOps.Horizontal.isMovable = true;
geoOps.Horizontal.initialize = function (el) {
    var pos = geoOps._helper.initializePoint(el);
    putStateComplexVector(pos);
};
geoOps.Horizontal.getParamForInput = function (el, pos, type) {
    var homog = List.cross(pos, List.ex);
    return List.normalizeMax(homog);
};
geoOps.Horizontal.getParamFromState = function (el) {
    return getStateComplexVector(3);
};
geoOps.Horizontal.putParamToState = function (el, param) {
    putStateComplexVector(param);
};
geoOps.Horizontal.updatePosition = function (el) {
    var param = getStateComplexVector(3);
    putStateComplexVector(param); // copy param
    el.homog = General.withUsage(param, "Line");
};
geoOps.Horizontal.stateSize = 6;
```

### geoOps.Vertical

```javascript
geoOps.Vertical = {};
geoOps.Vertical.kind = "L";
geoOps.Vertical.isMovable = true;
geoOps.Vertical.initialize = function (el) {
    var pos = geoOps._helper.initializePoint(el);
    putStateComplexVector(pos);
};
geoOps.Vertical.getParamForInput = function (el, pos, type) {
    var homog = List.cross(pos, List.ey);
    return List.normalizeMax(homog);
};
geoOps.Vertical.getParamFromState = function (el) {
    return getStateComplexVector(3);
};
geoOps.Vertical.putParamToState = function (el, param) {
    putStateComplexVector(param);
};
geoOps.Vertical.updatePosition = function (el) {
    var param = getStateComplexVector(3);
    putStateComplexVector(param); // copy param
    el.homog = General.withUsage(param, "Line");
};
geoOps.Vertical.stateSize = 6;
```

**Comment by gagern** [comment]
Hello @elkins0, welcome to the CindyJS open source project.

I had already noticed the inconsistencies you mention, and wrote about them on the developer's mailing list. So far we haven't reached a decision whether we should keep the current versions of `Vertical` and `Horizontal` or whether we should simply drop them. As you correctly said, the versions exported from Cinderella have different semantics. So I'd add the Cinderella semantics using Cinderella's names, even if doing so might lead to some confusion for those reading the code and finding two very similar operations. A comment somewhere might help in that respect.

The existing `FreeLine` is likely the result of an incomplete Copy&Paste&Edit operation, with edits to the name but not the body. It's definitely incorrect the way it is. That's why this report should be considered a bug report.

Would you be interested in creating two GitHub pull requests for your proposed code, one for `FreeLine` and the other for `HorizontalLine` and `VerticalLine`? That would be very useful, not only since it would make it easier for me to test integrate the code, but also because we could work out what information we need to provide in order to help new contributors, and because it would be useful in case you have further contributions in the future. The general description at https://help.github.com/articles/fork-a-repo/ should be a good starting point. https://github.com/CindyJS/CindyJS#contributing already states that new features should be accompanied by a test case, in this case one exported from Cinderella (but with the URL adjusted to refer to the local build).

**Comment by elkins0** [comment]
I tried using the web interface like I did on issue #78 here but get a "Secure Connection Failed" when I try to propose a change. I cannot even successfully preview. My usual connection is a dial-up and so very slow, perhaps that is why this happens on this five times larger file (9 KB versus 49 KB). I'll see if I have better luck using `GitHubSetup.exe`

Also is `src="../build/js/Cindy.js"` the right URL and should I somehow add the html file to the forked repo, perhaps under examples? I have been unable to attach cdy files to messages and fear the same would be true for html files.
